source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+  }
+  absl::flat_hash_map<std::string, MetricsPluginPtr> plugin_from_name;
+  std::vector<std::string> plugin_names = {"nighthawk.builtin"};
+  plugin_from_name["nighthawk.builtin"] =
+      std::make_unique<NighthawkStatsEmulatedMetricsPlugin>(nighthawk::client::Output());
+  for (const envoy::config::core::v3::TypedExtensionConfig& config :
+       spec.metrics_plugin_configs()) {
+    plugin_names.push_back(config.name());
+    absl::StatusOr<MetricsPluginPtr> metrics_plugin_or = LoadMetricsPlugin(config);
+    if (!metrics_plugin_or.ok()) {
+      errors.emplace_back(
+          absl::StrCat("Failed to load MetricsPlugin: ", metrics_plugin_or.status().message()));
+      continue;
+    }
+    plugin_from_name[config.name()] = std::move(metrics_plugin_or.value());
+  }
+  absl::StatusOr<StepControllerPtr> step_controller_or =
+      LoadStepControllerPlugin(spec.step_controller_config(), spec.nighthawk_traffic_template());
+  if (!step_controller_or.ok()) {
+    errors.emplace_back(absl::StrCat("Failed to load StepController plugin: ",
+                                     step_controller_or.status().message()));
+  }
+  std::vector<MetricSpec> all_metric_specs;
+  for (const MetricSpecWithThreshold& metric_threshold : spec.metric_thresholds()) {
+    all_metric_specs.push_back(metric_threshold.metric_spec());
+    absl::StatusOr<ScoringFunctionPtr> scoring_function_or =
+        LoadScoringFunctionPlugin(metric_threshold.threshold_spec().scoring_function());
+    if (!scoring_function_or.ok()) {
+      errors.emplace_back(absl::StrCat("Failed to load ScoringFunction plugin: ",
+                                       scoring_function_or.status().message()));
+    }
+  }
+  for (const MetricSpec& metric_spec : spec.informational_metric_specs()) {
+    all_metric_specs.push_back(metric_spec);
+  }
+  for (const MetricSpec& metric_spec : all_metric_specs) {
+    if (plugin_from_name.contains(metric_spec.metrics_plugin_name())) {
+      std::vector<std::string> supported_metrics =
+          plugin_from_name[metric_spec.metrics_plugin_name()]->GetAllSupportedMetricNames();
+      if (std::find(supported_metrics.begin(), supported_metrics.end(),
+                    metric_spec.metric_name()) == supported_metrics.end()) {
+        errors.emplace_back(
+            absl::StrCat("Metric named '", metric_spec.metric_name(),
+                         "' not implemented by plugin '", metric_spec.metrics_plugin_name(),
+                         "'. Metrics implemented: ", absl::StrJoin(supported_metrics, ", "), "."));
+      }
+    } else {
+      errors.emplace_back(absl::StrCat(
+          "MetricSpec referred to nonexistent metrics_plugin_name '",
+          metric_spec.metrics_plugin_name(),
+          "'. You must declare the plugin in metrics_plugin_configs or use plugin ",
+          "'nighthawk.builtin'. Available plugins: ", absl::StrJoin(plugin_names, ", "), "."));
+    }
+  }
+  if (errors.size() > 0) {
```

Comment:
I'm not sure how large this can grow; but if it hypothetically can grow very large, it might be a nice
courtesy to bound this somehow, possibly both in the input to avoid OOM and on the output to avoid spam.

Reply:
This should be bounded by the total number of fields in user-defined config proto. Each field can only result in at most one error line, as far as I know.